### DISCUSSION TITLE
How to set properly the UserDataFolder in WPF

### DISCUSSION THREAD

**Discussion by RicardPG-Vintra** [MicrosoftEdge/WebView2Feedback#4018] [Q&A]
I'm trying to set up the WebView2 in my project. I'm creating a dll that its going to be emmbeded on another APP (C#/WPF) as a "plugin". For permissions reason I'm getting an error, and it's loading a black screen.

![image](https://user-images.githubusercontent.com/23196043/216962454-27ff4cf5-2027-4f97-8d99-bbca66d64c68.png)

So, I need to specify the UserDataFolder.. after checking how to do it I fond this:
- https://learn.microsoft.com/en-us/microsoft-edge/webview2/concepts/user-data-folder?tabs=dotnet#specifying-a-custom-udf-location
- https://learn.microsoft.com/en-us/dotnet/api/microsoft.web.webview2.core.corewebview2environment.createasync?view=webview2-dotnet-1.0.864.35
- https://learn.microsoft.com/en-us/dotnet/api/microsoft.web.webview2.wpf.webview2.creationproperties?view=webview2-dotnet-1.0.1518.46#microsoft-web-webview2-wpf-webview2-creationproperties

None of the proposed ways from docs seems not taking any effect. Here my current code:

```
string dataFolder = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData) + @"\myAppData";
await CoreWebView2Environment.CreateAsync("", dataFolder);
MessageBox.Show(
  "After set folder: " + 
  CoreWebView2Environment.GetAvailableBrowserVersionString().ToString() + " " +
  dataFolder
); <-- It's getting correctly the version.
webView2 = new WebView2();
// webView2.CreationProperties.UserDataFolder = dataFolder; <-- This thow an error because NullException
// try {
//   CoreWebView2Environment environment = await CoreWebView2Environment.CreateAsync("", dataFolder);
//   await webView2.EnsureCoreWebView2Async(environment); <-- This wait forever, never ends
// }
//   catch (Exception ex)
// {
webView2.Loaded += WebView_OnLoaded;
webView2.CoreWebView2InitializationCompleted += WebView2OnCoreWebView2InitializationCompleted;
webView2.Source = new Uri("https://google.com");
BrowserContainer.Children.Add(webView2);
```

Also reviewed the https://github.com/MicrosoftEdge/WebView2Samples/tree/1710d535e895ed9c24196c5482e990e10d7285c7/SampleApps/WebView2WpfBrowser repo, and nothing related there.

I would appreciate if someone could provide a way to configure properly the Edge using WebView2.

Thanks!

NOTE: I'm using WebView2 v1.0.1518.46

**Comment by novac42** [comment]
@RicardPG-Vintra Thanks for reaching out. May you also let us know which version of operating system you are using? And the runtime version is also v1.0.1518.46?

**Comment by RicardPG-Vintra** [comment]
Sure. I'm using Windows 10 pro.

For the runtime thing, is quite dark for me.. I have 109.0.1518.78 installed yesterday for the WebView2 Runtime.

[image](https://user-images.githubusercontent.com/23196043/217208174-2640453c-15fd-4d15-9fc3-a91c85bdda18.png)

I installed the latest version available of the canary (Versión 112.0.1667.0). But, the  `CoreWebView2Environment.GetAvailableBrowserVersionString()` is returning 109.0.1518.78, wich is the version that, after check it, seems to be the version that I had Edge in the system. The v1.0.1518.46 that I was refering it's for the WebView2 package in C#, it's the most recent version available in nuget.

![image](https://user-images.githubusercontent.com/23196043/217208661-756abe10-b0ef-4ce2-b2df-b84ec274854f.png)

Thanks!

**Comment by sln162** [comment]
@RicardPG-Vintra    

> I installed the latest version available of the canary (Versión 112.0.1667.0). But, the CoreWebView2Environment.GetAvailableBrowserVersionString() is returning 109.0.1518.78

The default channel search order is the WebView2 Runtime, Beta, Dev, and Canary. When an override WEBVIEW2_RELEASE_CHANNEL_PREFERENCE environment variable or applicable releaseChannelPreference registry value is set to 1, the channel search order is reversed.

`Environment.SetEnvironmentVariable("WEBVIEW2_RELEASE_CHANNEL_PREFERENCE", "1");`

This code should be added before GetAvailableBrowserVersionString.
